src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,509 @@
+        #       `get_per_client_data()`
+        #   - if you were accidentally inconsistent with their values
+        #       then you would hit trouble quickly!
+        self.experiment_slug = experiment_slug
+        self.start_date = start_date
+        self.num_dates_enrollment = num_dates_enrollment
+        self.addon_version = addon_version
+
+    def get_enrollments(self, spark, study_type='pref_flip', end_date=None):
+        """Return a DataFrame of enrolled clients.
+
+        This works for pref-flip and addon studies.
+
+        The underlying queries are different for pref-flip vs addon
+        studies, because as of 2019/04/02, branch information isn't
+        reliably available in the `events` table for addon experiments:
+        branch may be NULL for all enrollments. The enrollment
+        information for them is most reliably available in
+        `telemetry_shield_study_parquet`. Once this issue is resolved,
+        we will probably start using normandy events for all desktop
+        studies.
+        Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1536644
+
+        Args:
+            spark: The spark context.
+            study_type (str): One of the following strings:
+                - 'pref_flip'
+                - 'addon'
+            end_date (str, optional): Ignore enrollments after this
+                date: for faster queries on stale experiments. If you
+                set `num_dates_enrollment` then do not set this; at best
+                it would be redundant, at worst it's contradictory.
+
+        Returns:
+            A Spark DataFrame of enrollment data. One row per
+            enrollment. Columns:
+                - client_id (str)
+                - enrollment_date (str): e.g. '20190329'
+                - branch (str)
+        """
+        if study_type == 'pref_flip':
+            enrollments = self._get_enrollments_view_normandy(spark)
+
+        elif study_type == 'addon':
+            enrollments = self._get_enrollments_view_addon(spark, self.addon_version)
+
+        # elif study_type == 'glean':
+        #     raise NotImplementedError
+
+        else:
+            raise ValueError("Unrecognized study_type: {}".format(study_type))
+
+        enrollments = enrollments.filter(
+            enrollments.enrollment_date >= self.start_date
+        ).filter(
+            enrollments.experiment_slug == self.experiment_slug
+        )
+
+        if self.num_dates_enrollment is not None:
+            assert end_date is None, "Possibly contradictory instructions!"
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= self._get_scheduled_max_enrollment_date()
+            )
+        elif end_date is not None:
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= end_date
+            )
+
+        enrollments.cache()
+
+        return enrollments
+
+    def get_per_client_data(
```

Comment:
This seems reasonable, but for my own context, why not generate enrollments automatically?

Reply:
If you're looking at a conversion window with `conv_window_start_days > 0` then you can't guarantee that `df` has a row for each enrolled user (e.g. if the user didn't send a ping during the conversion window because they had already churned). So you already need to query enrollments then left join onto the data. Given that you have to do this, the `events` table is a better source of information because compared to the main ping there's less latency between enrolling and the event being sent, which means less chance for people to enroll without us learning about them.

Reply:
I think I'm confused. I understand why this function depends on a table of enrollments, but I'm interested in why we're passing this function a table of enrollments instead of calling `self.get_enrollments()` inside the function. This seems like a flexible way to build the API (which I like and agree with) but it does require some additional config and offers a chance to accidentally break `enrollments` before passing to `get_per_client_data`. 

Like I said, this seems reasonable as is. I'm mostly making sure I understand what this function is doing and that I'm not missing some nuance.

Reply:
Yes, this is here for flexibility. E.g. this week Megan needed to re-analyze an old mobile experiment, and was able to build her own `enrollments` dataframe and feed that in.

Eventually I think we should support segmenting too, and I think that would involve filtering the `enrollments` dataframe before feeding it into `get_per_client_data`.

Also, if people want to query multiple data sources (e.g. I needed to query `main_summary` for some metrics but could get away with `clients_daily` for most) then they can reuse the one `enrollments` dataframe.

Likewise if people want data from multiple conversion windows (e.g. to build a time series) they can reuse `enrollments` - though in the future we'll probably have a method separate to `get_per_client_data` to solve that need.

When people don't need flexibility then they'll be able to use mozreport to drive all this

Reply:
Awesome, makes a lot of sense!